### ISSUE TITLE
RollTables seem to reference null IDs, result in broken links

### ISSUE THREAD

**Issue by wyrmisis** [CelestianGC/osric-compendium#1]
Hello! I'm posting this on behalf of a friend who's looking to start using OSRIC in Foundry.

While experimenting with using this set of compendia, he noticed that RollTables that reference another document in the compendia have broken links. On closer inspection, the compendium links on each item look to refer to null values.

Any thoughts on what's going on? I'll attach his exported chat log below.

```
[10/15/2022, 3:21:44 AM] Gamemaster
NPC Parties and Dragon encounters won't output to encounter boxes as other entries do, because they contain additional information and are packed in story entries. If you get what seems to be "no result", check the die result in chat. NPC (and Dragon, if appearing) tables are attached in the notes tab of this table to roll on separately.
1d200
1d200 57 57
57
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
Expand
fvtt-log-Sat-Oct-15-2022.txt
5 KB
﻿
[10/15/2022, 3:21:44 AM] Gamemaster
NPC Parties and Dragon encounters won't output to encounter boxes as other entries do, because they contain additional information and are packed in story entries. If you get what seems to be "no result", check the die result in chat. NPC (and Dragon, if appearing) tables are attached in the notes tab of this table to roll on separately.
1d200
1d200 57 57
57
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
@Compendium[world.osric-gm-table.null]{Monster Level 06 Otyugh, Lesser (type)}
1d200 = 57 = 57
---------------------------
[10/15/2022, 3:24:44 AM] Gamemaster
NPC Parties and Dragon encounters won't output to encounter boxes as other entries do, because they contain additional information and are packed in story entries. If you get what seems to be "no result", check the die result in chat. NPC (and Dragon, if appearing) tables are attached in the notes tab of this table to roll on separately.
1d200
1d200 68 68
68
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
@Compendium[osric.ruleset-documentation.null]{}
1d200 = 68 = 68
---------------------------
[10/15/2022, 3:24:48 AM] Gamemaster
NPC Parties and Dragon encounters won't output to encounter boxes as other entries do, because they contain additional information and are packed in story entries. If you get what seems to be "no result", check the die result in chat. NPC (and Dragon, if appearing) tables are attached in the notes tab of this table to roll on separately.
1d200
1d200 26 26
26
@Compendium[osric.ruleset-documentation.null]{}
1d200 = 26 = 26
---------------------------
[10/15/2022, 3:26:06 AM] Gamemaster
1d40
1d40 40 40
40
@Compendium[world.osric-gm-table.null]{Xorn - Number Appearing and "In Lair"}
1d40 = 40 = 40
---------------------------
[10/15/2022, 3:27:16 AM] Gamemaster
1d200
1d200 58 58
58
1 xorn are encountered away from their lair
@Compendium[world.osric-gm-table.null]{Xorn}
1d200 = 58 = 58
---------------------------
[10/15/2022, 3:30:44 AM] Gamemaster
1d200
1d200 85 85
85
3 xorn are encountered away from their lair
@Compendium[osric-compendium.osric-gm-actor.xUGJvV6TA4HFTUtO]{Xorn}
1d200 = 85 = 85
---------------------------
[10/15/2022, 3:38:36 AM] Xorn #6
Claw #1
Attack
Damage (d3/slashing)
---------------------------
[10/15/2022, 3:38:40 AM] Xorn #6
Damage wizard
1 Slashing damage
1d3 1 1
Undo
---------------------------
[10/15/2022, 4:02:15 AM] Xorn #6
NPC Parties and Dragon encounters won't output to encounter boxes as other entries do, because they contain additional information and are packed in story entries. If you get what seems to be "no result", check the die result in chat. NPC (and Dragon, if appearing) tables are attached in the notes tab of this table to roll on separately.
1d200
1d200 18 18
18
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
@Compendium[world.osric-gm-table.null]{Mummy (1d4)}
1d200 = 18 = 18
```

**Comment by 69bossmustang** [comment]
> Hello! I'm posting this on behalf of a friend who's looking to start using OSRIC in Foundry.
> 
> While experimenting with using this set of compendia, he noticed that RollTables that reference another document in the compendia have broken links. On closer inspection, the compendium links on each item look to refer to null values.
> 
> ```

Currently we're not focusing on the Table work in the OSRIC content for a few reasons. The biggest reason is in the OSRIC tables presume you can roll down multiple tables to get a final result. Foundry however only lets you nest 2 tables. Which makes doing any even a simple of 3 tables deep doesnt work and will report as bad links. The other reason is the tables are massive, there are a LOT and we're trying to focus on actors/items initially. I will bring this up with EOTB (he's working on the OSRIC content) to see if he can address something that "should" work (short table depth) but it probably wont be until after the ruleset beta has opened before he'll be able to look at it.

**Comment by EOTB1E** [comment]
It looks like these are rolls on the dungeon random encounter tables.  As Celestian mentioned, we postponed working on tables as soon as we found Foundry doesn't currently support more than very shallowly nested tables.  These tables are part of table nests that can go 15+ tables deep, depending on the result.

It looks like the specific null results are for what would be the Foundry actor linked to that table result.  But these are not yet linked (and so returning null results) because until the nesting problem is fixed, I wouldn't have any way to test the entire string of tables to ensure no linking mistakes were made.

tl;dr, these will be fixed when nesting works, but not before.

**Comment by wyrmisis** [comment]
Okay, cool, good to know it's a known thing.

If I get a chance, I'll see if I can tinker with extending Foundry's `RollTable` class to get it to handle deeply nested tables; a quick look at it suggests that `RollTable.roll()` isn't passing its `recursive` prop to subsequent rolls.

**Comment by EOTB1E** [comment]
Getting nested tables working right would be awesome!  So much prep power and ease-of-sandbox play is in those deeply nested tables.  Can roll up an NPC adventuring party in a click, etc., with only minor tweaking instead of time spent from a scratch state.

**Comment by 69bossmustang** [comment]
> Okay, cool, good to know it's a known thing.
> 
> If I get a chance, I'll see if I can tinker with extending Foundry's `RollTable` class to get it to handle deeply nested tables; a quick look at it suggests that `RollTable.roll()` isn't passing its `recursive` prop to subsequent rolls.

I've not looked deeply into this as I've been working on other things but this looks like the blocker.

in rol()...

```
    if ( _depth > 5 ) {
      throw new Error(`Maximum recursion depth exceeded when attempting to draw from RollTable ${this.id}`);
    }
```

**Comment by wyrmisis** [comment]
I saw that @CelestianGC  made the `RollTable` override in the OSRIC system today -- quick turnaround :)

I gave it a test and noticed that many of the RollTables were trying to use `world.osric-gm-table` in their `documentCollection` field; this compendium doesn't exist, and as a result, they're defaulting to `osric.ruleset-documentation`, as it's the first compendium in `game.packs`.

I was able to fix it with the following command in Linux, run from `osric-compendium/packs`:

```sh
find . -type f -exec sed -i 's/"world\./"osric-compendium./g' {} \; 
```

That'll replace every instance of `"world.` with `"osric-compendium.`

I haven't tested every table to know if this will 100% fix everything (good lord are there a lot of tables), but this worked when trying to generate a few parties of NPCs.

**Comment by 69bossmustang** [comment]
I did some tinkering over the past couple of days and was able to get the depth issue resolved (that was easy) but we had some table record issues that were a bit more complicated to resolve but I ran some tests with EOTB and we think they are clean now. We've got some long term plans to make the table system better but for now I think what we've got will work for initial version.

Thanks for the feedback, motivated me to look at this.
